File: src/main/java/org/example/kalkulationsprogramm/service/BuchungssatzAbleitung.java
package org.example.kalkulationsprogramm.service;

import org.example.kalkulationsprogramm.domain.Beleg;
import org.example.kalkulationsprogramm.domain.BelegKategorie;
import org.example.kalkulationsprogramm.domain.Sachkonto;

/**
 * Doppik-Variante A: das Beleg-Modell behaelt genau EIN {@code Sachkonto},
 * die Gegen-Seite wird deterministisch aus {@code BelegKategorie + Sachkonto}
 * abgeleitet. So bleibt der Datenstand schlank, der Kassenbuch-PDF-Export kann
 * trotzdem klassisch Soll/Haben darstellen, wie es der Steuerberater erwartet.
 *
 * <p>Mapping (Issue #61):</p>
 * <table>
 *   <tr><th>Kategorie</th><th>Soll</th><th>Haben</th></tr>
 *   <tr><td>KASSE_EINNAHME + Ertrag-Konto</td><td>Kasse</td><td>Sachkonto</td></tr>
 *   <tr><td>KASSE_EINNAHME ohne Sachkonto</td><td>Kasse</td><td>Bank</td></tr>
 *   <tr><td>KASSE_AUSGABE + Aufwand-Konto</td><td>Sachkonto</td><td>Kasse</td></tr>
 *   <tr><td>KASSE_AUSGABE ohne Sachkonto</td><td>?</td><td>Kasse</td></tr>
 *   <tr><td>PRIVATEINLAGE</td><td>Kasse</td><td>Privateinlage</td></tr>
 *   <tr><td>PRIVATENTNAHME</td><td>Privatentnahme</td><td>Kasse</td></tr>
 * </table>
 *
 * <p>Robust: wenn ein Sachkonto fehlt, steht "?" statt einer NPE. Der Buchhalter
 * sieht im PDF sofort, wo nachgepflegt werden muss.</p>
 */
public final class BuchungssatzAbleitung {

    /** Label fuer die Kasse-Seite. Nicht "1600 Kasse" — der Steuerberater
     *  bekommt ohnehin den eigenen Kontenrahmen; im Kassenbuch zaehlt Klarheit. */
    public static final String KASSE = "Kasse";
    public static final String BANK = "Bank";
    public static final String PRIVATEINLAGE = "Privateinlage";
    public static final String PRIVATENTNAHME = "Privatentnahme";
    public static final String UNKLAR = "?";

    private BuchungssatzAbleitung() {
        // Utility-Klasse
    }

    /**
     * Liefert Soll/Haben-Labels fuer einen Beleg. Belege ohne Kassen-Bewegung
     * (BANK, KREDITKARTE, SONSTIGER_BELEG, UNZUGEORDNET) bekommen {@link #UNKLAR}
     * auf beiden Seiten — der Aufrufer entscheidet, ob er sie ueberhaupt
     * rendert.
     */
    public static Buchungssatz ableiten(Beleg beleg) {
        if (beleg == null || beleg.getBelegKategorie() == null) {
            return new Buchungssatz(UNKLAR, UNKLAR);
        }
        BelegKategorie kategorie = beleg.getBelegKategorie();
        String sachkontoLabel = sachkontoLabel(beleg.getSachkonto());

        return switch (kategorie) {
            case KASSE_EINNAHME -> new Buchungssatz(
                    KASSE,
                    beleg.getSachkonto() != null ? sachkontoLabel : BANK);
            case KASSE_AUSGABE -> new Buchungssatz(
                    beleg.getSachkonto() != null ? sachkontoLabel : UNKLAR,
                    KASSE);
            case PRIVATEINLAGE -> new Buchungssatz(
                    KASSE,
                    beleg.getSachkonto() != null ? sachkontoLabel : PRIVATEINLAGE);
            case PRIVATENTNAHME -> new Buchungssatz(
                    beleg.getSachkonto() != null ? sachkontoLabel : PRIVATENTNAHME,
                    KASSE);
            // Nicht-Kassen-Belege landen ueblicherweise nicht im Kassenbuch-Bereich
            // des PDFs — falls doch, kommt UNKLAR sichtbar an.
            case BANK, KREDITKARTE, SONSTIGER_BELEG, UNZUGEORDNET ->
                    new Buchungssatz(UNKLAR, UNKLAR);
        };
    }

    private static String sachkontoLabel(Sachkonto sk) {
        if (sk == null) return UNKLAR;
        String nr = sk.getNummer();
        String bz = sk.getBezeichnung();
        if (nr != null && !nr.isBlank() && bz != null && !bz.isBlank()) {
            return nr + " " + bz;
        }
        if (bz != null && !bz.isBlank()) return bz;
        if (nr != null && !nr.isBlank()) return nr;
        return UNKLAR;
    }

    /**
     * Einfaches Ergebnis-Record. Beide Seiten sind nie {@code null} —
     * fehlende Information wird als {@link #UNKLAR} ("?") signalisiert.
     */
    public record Buchungssatz(String soll, String haben) { }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/BuchungssatzAbleitung.kt
package org.example.kalkulationsprogramm.service

import org.example.kalkulationsprogramm.domain.Beleg
import org.example.kalkulationsprogramm.domain.BelegKategorie
import org.example.kalkulationsprogramm.domain.Sachkonto

object BuchungssatzAbleitung {
    const val KASSE = "Kasse"
    const val BANK = "Bank"
    const val PRIVATEINLAGE = "Privateinlage"
    const val PRIVATENTNAHME = "Privatentnahme"
    const val UNKLAR = "?"

    @JvmStatic
    fun ableiten(beleg: Beleg?): Buchungssatz {
        val kategorie = readField(beleg, "belegKategorie") as? BelegKategorie
            ?: return Buchungssatz(UNKLAR, UNKLAR)
        val sachkonto = readField(beleg, "sachkonto") as? Sachkonto
        val sachkontoLabel = sachkontoLabel(sachkonto)

        return when (kategorie) {
            BelegKategorie.KASSE_EINNAHME -> Buchungssatz(
                KASSE,
                if (sachkonto != null) sachkontoLabel else BANK,
            )
            BelegKategorie.KASSE_AUSGABE -> Buchungssatz(
                if (sachkonto != null) sachkontoLabel else UNKLAR,
                KASSE,
            )
            BelegKategorie.PRIVATEINLAGE -> Buchungssatz(
                KASSE,
                if (sachkonto != null) sachkontoLabel else PRIVATEINLAGE,
            )
            BelegKategorie.PRIVATENTNAHME -> Buchungssatz(
                if (sachkonto != null) sachkontoLabel else PRIVATENTNAHME,
                KASSE,
            )
            BelegKategorie.BANK,
            BelegKategorie.KREDITKARTE,
            BelegKategorie.SONSTIGER_BELEG,
            BelegKategorie.UNZUGEORDNET,
            -> Buchungssatz(UNKLAR, UNKLAR)
        }
    }

    private fun sachkontoLabel(sk: Sachkonto?): String {
        if (sk == null) return UNKLAR
        val nummer = readField(sk, "nummer") as? String
        val bezeichnung = readField(sk, "bezeichnung") as? String
        if (!nummer.isNullOrBlank() && !bezeichnung.isNullOrBlank()) {
            return "$nummer $bezeichnung"
        }
        if (!bezeichnung.isNullOrBlank()) return bezeichnung
        if (!nummer.isNullOrBlank()) return nummer
        return UNKLAR
    }

    private fun readField(target: Any?, fieldName: String): Any? {
        if (target == null) return null
        var type: Class<*>? = target.javaClass
        while (type != null) {
            try {
                val field = type.getDeclaredField(fieldName)
                field.isAccessible = true
                return field.get(target)
            } catch (_: NoSuchFieldException) {
                type = type.superclass
            }
        }
        return null
    }

    data class Buchungssatz(
        val soll: String,
        val haben: String,
    )
}
